File: Platforms/Windows/Reign.Video.D3D9/ShaderVariable.cpp
#include "pch.h"
#include "PixelShader.h"
#include "Shader.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region
	string^ ShaderVariable::Name::get() {return name;}
	#pragma endregion

	#pragma region Constructors
	ShaderVariable::ShaderVariable(VideoI^ video, D3DXHANDLE vertexHandle, D3DXHANDLE pixelHandle, ID3DXConstantTable* vertexVariables, ID3DXConstantTable* pixelVariables, string^ name)
	{
		this->video = (Video^)video;
		this->vertexHandle = vertexHandle;
		this->pixelHandle = pixelHandle;
		this->vertexVariables = vertexVariables;
		this->pixelVariables = pixelVariables;

		this->name = name;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setNothing);
		valueArrayObject = gcnew WeakReference(nullptr);
	}
	#pragma endregion

	#pragma region Methods
	void ShaderVariable::setNothing()
	{
		// Place holder.
	}

	void ShaderVariable::setFloat()
	{
		if (vertexHandle) vertexVariables->SetFloat(video->Device, vertexHandle, valueObject.X);
		if (pixelHandle) pixelVariables->SetFloat(video->Device, pixelHandle, valueObject.X);
	}

	void ShaderVariable::setVector2()
	{
		D3DXVECTOR4 vector;
		vector.x = valueObject.X;
		vector.y = valueObject.Y;
		vector.z = 0;
		vector.w = 0;
		if (vertexHandle) vertexVariables->SetVector(video->Device, vertexHandle, &vector);
		if (pixelHandle) pixelVariables->SetVector(video->Device, pixelHandle, &vector);
	}

	void ShaderVariable::setVector3()
	{
		D3DXVECTOR4 vector;
		vector.x = valueObject.X;
		vector.y = valueObject.Y;
		vector.z = valueObject.Z;
		vector.w = 0;
		if (vertexHandle) vertexVariables->SetVector(video->Device, vertexHandle, &vector);
		if (pixelHandle) pixelVariables->SetVector(video->Device, pixelHandle, &vector);
	}

	void ShaderVariable::setVector4()
	{
		pin_ptr<void> valuePtr = &valueObject.Vector4;
		D3DXVECTOR4* ptr = (D3DXVECTOR4*)(void*)valuePtr;
		if (vertexHandle) vertexVariables->SetVector(video->Device, vertexHandle, ptr);
		if (pixelHandle) pixelVariables->SetVector(video->Device, pixelHandle, ptr);
	}

	void ShaderVariable::setMatrix2()
	{
		// use set raw value
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::setMatrix3()
	{
		// use set raw value
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::setMatrix4()
	{
		pin_ptr<void> valuePtr = &valueObject.Matrix4;
		D3DXMATRIX* ptr = (D3DXMATRIX*)(void*)valuePtr;
		if (vertexHandle) vertexVariables->SetMatrix(video->Device, vertexHandle, ptr);
		if (pixelHandle) pixelVariables->SetMatrix(video->Device, pixelHandle, ptr);
	}

	void ShaderVariable::setFloatArray()
	{
		array<float>^ value = (array<float>^)valueArrayObject->Target;
		pin_ptr<void> valuePtr = &value[valueArrayOffset];
		float* ptr = ((float*)(void*)valuePtr) + valueArrayOffset;
		if (vertexHandle) vertexVariables->SetFloatArray(video->Device, vertexHandle, ptr, valueArrayCount);
		if (pixelHandle) pixelVariables->SetFloatArray(video->Device, pixelHandle, ptr, valueArrayCount);
	}

	void ShaderVariable::setVector4Array()
	{
		array<Vector4>^ value = (array<Vector4>^)valueArrayObject->Target;
		pin_ptr<void> valuePtr = &value[valueArrayOffset];
		D3DXVECTOR4* ptr = ((D3DXVECTOR4*)(void*)valuePtr) + valueArrayOffset;
		if (vertexHandle) vertexVariables->SetVectorArray(video->Device, vertexHandle, ptr, valueArrayCount);
		if (pixelHandle) pixelVariables->SetVectorArray(video->Device, pixelHandle, ptr, valueArrayCount);
	}

	void ShaderVariable::setMatrix4Array()
	{
		array<Matrix4>^ value = (array<Matrix4>^)valueArrayObject->Target;
		pin_ptr<void> valuePtr = &value[valueArrayOffset];
		D3DXMATRIX* ptr = ((D3DXMATRIX*)(void*)valuePtr) + valueArrayOffset;
		if (vertexHandle) vertexVariables->SetMatrixArray(video->Device, vertexHandle, ptr, valueArrayCount);
		if (pixelHandle) pixelVariables->SetMatrixArray(video->Device, pixelHandle, ptr, valueArrayCount);
	}

	void ShaderVariable::Set(float value)
	{
		valueObject.X = value;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setFloat);
	}

	void ShaderVariable::Set(float x, float y)
	{
		valueObject.X = x;
		valueObject.Y = y;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector2);
	}

	void ShaderVariable::Set(float x, float y, float z)
	{
		valueObject.X = x;
		valueObject.Y = y;
		valueObject.Z = z;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector3);
	}

	void ShaderVariable::Set(float x, float y, float z, float w)
	{
		valueObject.X = x;
		valueObject.Y = y;
		valueObject.Z = z;
		valueObject.W = w;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector4);
	}

	void ShaderVariable::Set(Vector2 value)
	{
		valueObject.Vector2 = value;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector2);
	}

	void ShaderVariable::Set(Vector3 value)
	{
		valueObject.Vector3 = value;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector3);
	}

	void ShaderVariable::Set(Vector4 value)
	{
		valueObject.Vector4 = value;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector4);
	}

	void ShaderVariable::Set(Matrix2 value)
	{
		valueObject.Matrix2 = value;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setMatrix2);
	}

	void ShaderVariable::Set(Matrix3 value)
	{
		valueObject.Matrix3 = value;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setMatrix3);
	}

	void ShaderVariable::Set(Matrix4 value)
	{
		valueObject.Matrix4 = value;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setMatrix4);
	}

	void ShaderVariable::Set(array<float>^ values)
	{
		valueArrayOffset = 0;
		valueArrayCount = values->Length;
		valueArrayObject->Target = values;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setFloatArray);
	}

	void ShaderVariable::Set(array<Vector2>^ values)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Vector3>^ values)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Vector4>^ values)
	{
		valueArrayOffset = 0;
		valueArrayCount = values->Length;
		valueArrayObject->Target = values;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector4Array);
	}

	void ShaderVariable::Set(array<Matrix2>^ values)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Matrix3>^ values)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Matrix4>^ values)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<float>^ values, int count)
	{
		valueArrayOffset = 0;
		valueArrayCount = count;
		valueArrayObject->Target = values;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setFloatArray);
	}

	void ShaderVariable::Set(array<Vector2>^ values, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Vector3>^ values, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Vector4>^ values, int count)
	{
		valueArrayOffset = 0;
		valueArrayCount = count;
		valueArrayObject->Target = values;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector4Array);
	}

	void ShaderVariable::Set(array<Matrix2>^ values, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Matrix3>^ values, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Matrix4>^ values, int count)
	{
		valueArrayOffset = 0;
		valueArrayCount = count;
		valueArrayObject->Target = values;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setMatrix4Array);
	}

	void ShaderVariable::Set(array<float>^ values, int offset, int count)
	{
		valueArrayOffset = offset;
		valueArrayCount = count;
		valueArrayObject->Target = values;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setFloatArray);
	}

	void ShaderVariable::Set(array<Vector2>^ values, int offset, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Vector3>^ values, int offset, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Vector4>^ values, int offset, int count)
	{
		valueArrayOffset = offset;
		valueArrayCount = count;
		valueArrayObject->Target = values;
		Apply = gcnew ApplyFunc(this, &ShaderVariable::setVector4Array);
	}

	void ShaderVariable::Set(array<Matrix2>^ values, int offset, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Matrix3>^ values, int offset, int count)
	{
		throw gcnew NotImplementedException();
	}

	void ShaderVariable::Set(array<Matrix4>^ values, int offset, int count)
	{
		throw gcnew NotImplementedException();
	}
	#pragma endregion
}
}
}

File: Platforms/Common/Reign.Video.OpenGL/Shaders/ShaderVariable.cs
using System;
using Reign.Core;
using System.Runtime.InteropServices;

namespace Reign.Video.OpenGL
{
	public class ShaderVariable : ShaderVariableI
	{
		[StructLayout(LayoutKind.Explicit)]
		private struct ValueObject
		{
			[FieldOffset(0)]
			public float X;

			[FieldOffset(4)]
			public float Y;

			[FieldOffset(8)]
			public float Z;

			[FieldOffset(12)]
			public float W;

			[FieldOffset(0)]
			public Vector2 Vector2;

			[FieldOffset(0)]
			public Vector3 Vector3;

			[FieldOffset(0)]
			public Vector4 Vector4;

			[FieldOffset(0)]
			public Matrix2 Matrix2;

			[FieldOffset(0)]
			public Matrix3 Matrix3;

			[FieldOffset(0)]
			public Matrix4 Matrix4;
		}

		#region Properties
		internal delegate void ApplyFunc();
		internal ApplyFunc Apply;
		private ValueObject valueObject;
		private WeakReference valueArrayObject;
		private int valueArrayOffset, valueArrayCount;

		private int location;
		public string Name {get; private set;}
		#endregion

		#region Constructors
		public ShaderVariable(int location, string name)
		{
			this.location = location;
			this.Name = name;

			Apply = setNothing;
			valueArrayObject = new WeakReference(null);
		}
		#endregion

		#region Methods
		private void setNothing()
		{
			// Place holder.
		}

		private void setFloat()
		{
			GL.Uniform1f(location, valueObject.X);
		}

		private void setVector2()
		{
			GL.Uniform2f(location, valueObject.X, valueObject.Y);
		}

		private void setVector3()
		{
			GL.Uniform3f(location, valueObject.X, valueObject.Y, valueObject.Z);
		}

		private void setVector4()
		{
			GL.Uniform4f(location, valueObject.X, valueObject.Y, valueObject.Z, valueObject.W);
		}

		private unsafe void setMatrix2()
		{
			fixed (void* ptr = &valueObject.Matrix2)
			{
				GL.UniformMatrix2fv(location, 1, false, (float*)ptr);
			}
		}

		private unsafe void setMatrix3()
		{
			fixed (void* ptr = &valueObject.Matrix3)
			{
				GL.UniformMatrix3fv(location, 1, false, (float*)ptr);
			}
		}

		private unsafe void setMatrix4()
		{
			fixed (void* ptr = &valueObject.Matrix4)
			{
				GL.UniformMatrix4fv(location, 1, false, (float*)ptr);
			}
		}

		private unsafe void setFloatArray()
		{
			fixed (void* ptr = ((float[])valueArrayObject.Target))
			{
				GL.Uniform1fv(location, valueArrayCount, (float*)ptr + valueArrayOffset);
			}
		}

		private unsafe void setVector4Array()
		{
			fixed (void* ptr = ((Vector4[])valueArrayObject.Target))
			{
				GL.Uniform4fv(location, valueArrayCount, (float*)ptr + (valueArrayOffset * 4));
			}
		}

		private unsafe void setMatrix4Array()
		{
			fixed (void* ptr = ((Matrix4[])valueArrayObject.Target))
			{
				GL.UniformMatrix4fv(location, valueArrayCount, false, (float*)ptr + (valueArrayOffset * 12));
			}
		}

		public void Set(float value)
		{
			valueObject.X = value;
			Apply = setFloat;
		}

		public void Set(float x, float y)
		{
			valueObject.X = x;
			valueObject.Y = y;
			Apply = setVector2;
		}

		public void Set(float x, float y, float z)
		{
			valueObject.X = x;
			valueObject.Y = y;
			valueObject.Z = z;
			Apply = setVector3;
		}

		public void Set(float x, float y, float z, float w)
		{
			valueObject.X = x;
			valueObject.Y = y;
			valueObject.Z = z;
			valueObject.W = w;
			Apply = setVector4;
		}

		public void Set(Vector2 value)
		{
			valueObject.Vector2 = value;
			Apply = setVector2;
		}

		public void Set(Vector3 value)
		{
			valueObject.Vector3 = value;
			Apply = setVector3;
		}

		public void Set(Vector4 value)
		{
			valueObject.Vector4 = value;
			Apply = setVector4;
		}

		public void Set(Matrix2 value)
		{
			valueObject.Matrix2 = value;
			Apply = setMatrix2;
		}

		public void Set(Matrix3 value)
		{
			valueObject.Matrix3 = value;
			Apply = setMatrix3;
		}

		public void Set(Matrix4 value)
		{
			valueObject.Matrix4 = value;
			Apply = setMatrix4;
		}

		public void Set(float[] values)
		{
			valueArrayOffset = 0;
			valueArrayCount = values.Length;
			valueArrayObject.Target = values;
			Apply = setFloatArray;
		}

		public void Set(Vector2[] values)
		{
			throw new NotImplementedException();
		}

		public void Set(Vector3[] values)
		{
			throw new NotImplementedException();
		}

		public void Set(Vector4[] values)
		{
			valueArrayOffset = 0;
			valueArrayCount = values.Length;
			valueArrayObject.Target = values;
			Apply = setVector4Array;
		}

		public void Set(Matrix2[] values)
		{
			throw new NotImplementedException();
		}

		public void Set(Matrix3[] values)
		{
			throw new NotImplementedException();
		}

		public void Set(Matrix4[] values)
		{
			throw new NotImplementedException();
		}

		public void Set(float[] values, int count)
		{
			valueArrayOffset = 0;
			valueArrayCount = count;
			valueArrayObject.Target = values;
			Apply = setFloatArray;
		}

		public void Set(Vector2[] values, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Vector3[] values, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Vector4[] values, int count)
		{
			valueArrayOffset = 0;
			valueArrayCount = count;
			valueArrayObject.Target = values;
			Apply = setVector4Array;
		}

		public void Set(Matrix2[] values, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Matrix3[] values, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Matrix4[] values, int count)
		{
			valueArrayOffset = 0;
			valueArrayCount = count;
			valueArrayObject.Target = values;
			Apply = setMatrix4Array;
		}

		public void Set(float[] values, int offset, int count)
		{
			valueArrayOffset = offset;
			valueArrayCount = count;
			valueArrayObject.Target = values;
			Apply = setFloatArray;
		}

		public void Set(Vector2[] values, int offset, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Vector3[] values, int offset, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Vector4[] values, int offset, int count)
		{
			valueArrayOffset = offset;
			valueArrayCount = count;
			valueArrayObject.Target = values;
			Apply = setVector4Array;
		}

		public void Set(Matrix2[] values, int offset, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Matrix3[] values, int offset, int count)
		{
			throw new NotImplementedException();
		}

		public void Set(Matrix4[] values, int offset, int count)
		{
			throw new NotImplementedException();
		}
		#endregion
	}
}